interzeptor.go

```diff
@@ -0,0 +1,140 @@
+// enforcement of the services themselves.
+type Interzeptor struct {
+	acctMgr accounter
+
+	authz interzeptor
+	pathz interzeptor
+}
+
+// newInterzeptor constructs a new interceptor entity.
+func newInterzeptor(acctMgr accounter, authz interzeptor, pathz interzeptor) *Interzeptor {
+	return &Interzeptor{
+		acctMgr: acctMgr,
+		pathz:   pathz,
+		authz:   authz,
+	}
+}
+
+// UnaryInterceptor is a gNSI unary interceptor.
+func (i *Interzeptor) UnaryInterceptor(ctx context.Context, req any, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (any, error) {
+	if err := i.interzept(ctx, info.FullMethod); err != nil {
+		return nil, err
+	}
+
+	return handler(ctx, req)
+}
+
+// StreamInterceptor is a gNSI stream interceptor
+func (i *Interzeptor) StreamInterceptor(srv any, ss grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
+	if err := i.interzept(ss.Context(), info.FullMethod); err != nil {
+		return err
+	}
+
+	return handler(srv, ss)
+}
+
+// interzept executes all registered interceptors in parallel. It retrieves any accounting
+// information from each interceptor and records it. It finally returns any error such that
+// the gNSI service is enforced.
+func (i *Interzeptor) interzept(ctx context.Context, fullMethod string) error {
+	record, err := i.acctMgr.RecordFromContext(ctx)
+	defer i.acctMgr.Account(record)
+	if err != nil {
+		return err
+	}
+
+	if err := i.authz.Authorize(ctx); err != nil {
+		if status.Code(err) == codes.PermissionDenied {
+			record.ServiceRequest = &apb.RecordResponse_GrpcService{
+				GrpcService: &apb.GrpcService{
+					RpcName:     fullMethod,
+					ServiceType: apb.GrpcService_GRPC_SERVICE_TYPE_UNSPECIFIED, // because we are not one of the defined types.
+					Authz: &apb.AuthzDetail{
+						Status: apb.AuthzDetail_AUTHZ_STATUS_DENY,
+						Detail: "unauthorized by policy",
+					},
+				},
+			}
+			return status.Errorf(codes.PermissionDenied, "permission denied")
+		} else {
+			record.ServiceRequest = &apb.RecordResponse_GrpcService{
+				GrpcService: &apb.GrpcService{
+					RpcName:     fullMethod,
+					ServiceType: apb.GrpcService_GRPC_SERVICE_TYPE_UNSPECIFIED, // because we are not one of the defined types.
+					Authz: &apb.AuthzDetail{
+						Status: apb.AuthzDetail_AUTHZ_STATUS_ERROR,
+						Detail: err.Error(),
+					},
+				},
+			}
+			return status.Errorf(codes.PermissionDenied, "failed to evaluate authorization: %v", err)
+		}
+	}
+
+	record.ServiceRequest = &apb.RecordResponse_GrpcService{
+		GrpcService: &apb.GrpcService{
+			RpcName:     fullMethod,
+			ServiceType: apb.GrpcService_GRPC_SERVICE_TYPE_UNSPECIFIED, // because we are not one of the defined types.
+			Authz: &apb.AuthzDetail{
+				Status: apb.AuthzDetail_AUTHZ_STATUS_PERMIT,
+			},
+		},
+	}
+
+	// TODO(alshabib) handle pathz here.
+
+	return nil
+}
+
+// SetupInterzeptors constructs the interzeptors for gNSI services. It also performs accounting
```

Comment:
Is there any case where using the global above causes a problem? For example, if I have a single binary that exposes two different services that have their own policies, then doesn't this mean that we would need to call `SetupInterzeptors` more than once and hence overwrite the global?

Should we rather have something whereby we return a new instance of the interzeptor to the caller, to allow for concurrency?

Reply:
See comment above. This is going away.